File: services/api/routes/blog.py
"""
Blog endpoints:
  GET /blog          — paginated list of blog posts
  GET /blog/{slug}   — single blog post with joined event previews
"""
from __future__ import annotations

from db import get_client
from router import ok, error

_MAX_LIMIT = 50


# ── GET /blog ──────────────────────────────────────────────────────────────────

def list_blog_posts(event: dict, ctx) -> dict:
    qs        = event.get("queryStringParameters") or {}
    limit     = min(int(qs.get("limit", 12)), _MAX_LIMIT)
    offset    = int(qs.get("offset", 0))
    post_type = qs.get("post_type")
    area      = qs.get("area")

    db = get_client()

    count_q = db.table("blog_posts").select("id", count="exact")
    data_q  = (
        db.table("blog_posts")
        .select(
            "id, slug, title, meta_description, post_type, area, "
            "date_range_start, date_range_end, event_count, "
            "hero_image_url, published_at"
        )
        .order("published_at", desc=True)
        .range(offset, offset + limit - 1)
    )

    if post_type:
        count_q = count_q.eq("post_type", post_type)
        data_q  = data_q.eq("post_type", post_type)

    if area:
        count_q = count_q.eq("area", area)
        data_q  = data_q.eq("area", area)

    count_resp = count_q.execute()
    data_resp  = data_q.execute()

    total = count_resp.count or 0
    items = data_resp.data or []

    return ok({
        "items":    items,
        "total":    total,
        "limit":    limit,
        "offset":   offset,
        "has_more": (offset + limit) < total,
    })


# ── GET /blog/{slug} ───────────────────────────────────────────────────────────

def get_blog_post(event: dict, ctx) -> dict:
    slug = (event.get("pathParameters") or {}).get("slug", "")
    if not slug:
        return error("Missing slug", 400)

    db = get_client()

    # Fetch the blog post
    post_resp = (
        db.table("blog_posts")
        .select("*")
        .eq("slug", slug)
        .single()
        .execute()
    )

    if not post_resp.data:
        return error("Blog post not found", 404)

    post = post_resp.data

    # Fetch lightweight event previews for the event_ids in this post
    event_ids = post.get("event_ids") or []
    events    = []
    if event_ids:
        events_resp = (
            db.table("events")
            .select(
                "id, slug, title, start_at, venue_name, location_text, "
                "is_free, cost_description, tags, event_type, "
                "image_url_sm, image_alt, image_blurhash, image_width, image_height"
            )
            .in_("id", event_ids)
            .eq("status", "published")
            .order("start_at", desc=False)
            .execute()
        )
        events = events_resp.data or []

    post["events"] = events
    return ok(post)

File: apps/web/src/types/blog.ts
import type { Event } from './events'

export type PostType =
  | 'weekend_roundup'
  | 'week_ahead'
  | 'location_specific'
  | 'free_events'
  | 'indoor'
  | 'date_specific'

export type PostArea =
  | 'nova'
  | 'fairfax'
  | 'loudoun'
  | 'arlington'
  | 'alexandria'
  | 'prince_william'
  | 'dc_adjacent'

export interface BlogPost {
  id: string
  slug: string
  title: string
  meta_description: string | null
  content: string            // markdown
  post_type: PostType
  area: PostArea
  date_range_start: string   // ISO date YYYY-MM-DD
  date_range_end: string
  event_count: number
  event_ids: string[]
  hero_image_url: string | null
  published_at: string
}

export interface BlogPostWithEvents extends BlogPost {
  events: EventPreview[]
}

// Lightweight event shape returned with blog post detail
export interface EventPreview {
  id: string
  slug: string
  title: string
  start_at: string
  venue_name: string
  location_text: string | null
  is_free: boolean
  cost_description: string | null
  tags: string[]
  event_type: string
  image_url_sm: string | null
  image_alt: string | null
  image_blurhash: string | null
  image_width: number | null
  image_height: number | null
}

export interface BlogPostsResponse {
  items: BlogPost[]
  total: number
  limit: number
  offset: number
  has_more: boolean
}

export interface BlogPostsParams {
  limit?: number
  offset?: number
  post_type?: PostType
  area?: PostArea
}

// Human-readable labels for display
export const POST_TYPE_LABELS: Record<PostType, string> = {
  weekend_roundup:   'This Weekend',
  week_ahead:        'This Week',
  location_specific: 'By Location',
  free_events:       'Free Events',
  indoor:            'Indoor Activities',
  date_specific:     'Specific Date',
}

export const AREA_LABELS: Record<PostArea, string> = {
  nova:          'Northern Virginia',
  fairfax:       'Fairfax County',
  loudoun:       'Loudoun County',
  arlington:     'Arlington',
  alexandria:    'Alexandria',
  prince_william:'Prince William County',
  dc_adjacent:   'DC Area',
}
